File: MetaheuristicOptimization/AllMetaheuristic/others/CRO_re/CRO/CRO.m
% Coral Reefs Optimization (CRO) Implementation

% Parameters
population_size = 50;        % Population size
num_variables = 10;          % Number of variables
max_iterations = 100;        % Maximum number of iterations
num_reefs = 10;              % Number of reefs (subpopulations)
alpha = 0.1;                  % Learning rate
lower_limit = -5;            % Lower limit for variables
upper_limit = 5;             % Upper limit for variables

% Initialization
population = cell(num_reefs, 1);
for i = 1:num_reefs
    population{i} = lower_limit + (upper_limit - lower_limit) * rand(population_size, num_variables);  % Random initial population
end

% Main loop
for iter = 1:max_iterations
    % Evaluate fitness for each reef
    fitness = zeros(num_reefs, population_size);
    for i = 1:num_reefs
        for j = 1:population_size
            fitness(i, j) = objective_function(population{i}(j, :));
        end
    end
    
    % Exchange solutions between reefs (migration)
    for i = 1:num_reefs
        for j = 1:num_reefs
            if i ~= j
                % Randomly select a solution to migrate
                idx = randi([1, population_size]);
                solution_to_migrate = population{i}(idx, :);
                
                % Replace a random solution in the destination reef
                idx_replace = randi([1, population_size]);
                population{j}(idx_replace, :) = solution_to_migrate;
            end
        end
    end
    
    % Update each reef's population
    for i = 1:num_reefs
        for j = 1:population_size
            % Perform local search (e.g., gradient descent) on each solution
            % Here, we use a simple random perturbation
            population{i}(j, :) = population{i}(j, :) + alpha * randn(1, num_variables);
            
            % Ensure the new position is within bounds
            population{i}(j, :) = max(min(population{i}(j, :), upper_limit), lower_limit);
        end
    end
end

% Find the best solution among all reefs
best_fitness = inf;
best_solution = [];
for i = 1:num_reefs
    for j = 1:population_size
        if objective_function(population{i}(j, :)) < best_fitness
            best_fitness = objective_function(population{i}(j, :));
            best_solution = population{i}(j, :);
        end
    end
end

% Display best solution
fprintf('Best solution: %s\n', mat2str(best_solution));
fprintf('Best fitness: %f\n', best_fitness);
% Objective function (example)
function z = objective_function(x)
    % Example: Sphere function
    z = sum(x.^2);
end

File: chekml/MetaheuristicOptimization/CIntegration_ctypes/CRO.c
#include "CRO.h"
#include "generaloptimizer.h"
#include <stdlib.h>
#include <time.h>

// Function to generate a random double between min and max
double rand_double(double min, double max);

// Initialize reefs randomly within bounds
void initialize_reefs(Optimizer *opt) {
    if (!opt || !opt->bounds || !opt->population) {
        fprintf(stderr, "Invalid optimizer, bounds, or population\n");
        opt->population = NULL;
        return;
    }

    // Compute solutions per reef
    int solutions_per_reef = opt->population_size / NUM_REEFS;
    if (opt->population_size % NUM_REEFS != 0 || solutions_per_reef <= 0) {
        fprintf(stderr, "Population size %d must be divisible by NUM_REEFS %d\n", opt->population_size, NUM_REEFS);
        opt->population = NULL;
        return;
    }
    if (solutions_per_reef != POPULATION_SIZE) {
        fprintf(stderr, "Warning: Using %d solutions per reef instead of POPULATION_SIZE %d\n", solutions_per_reef, POPULATION_SIZE);
    }

    // Initialize positions within bounds
    for (int i = 0; i < NUM_REEFS; i++) {
        for (int j = 0; j < solutions_per_reef; j++) {
            int idx = i * solutions_per_reef + j;
            for (int k = 0; k < opt->dim; k++) {
                double lower = opt->bounds[2 * k];
                double upper = opt->bounds[2 * k + 1];
                opt->population[idx].position[k] = lower + (upper - lower) * rand_double(0.0, 1.0);
            }
            opt->population[idx].fitness = INFINITY;
        }
    }
    enforce_bound_constraints(opt);
}

// Evaluate fitness for each solution in each reef
void evaluate_reefs(Optimizer *opt, double (*objective_function)(double *)) {
    if (!opt || !opt->population || !objective_function) {
        fprintf(stderr, "Invalid optimizer, population, or objective function\n");
        return;
    }
    int solutions_per_reef = opt->population_size / NUM_REEFS;
    for (int i = 0; i < NUM_REEFS; i++) {
        for (int j = 0; j < solutions_per_reef; j++) {
            int idx = i * solutions_per_reef + j;
            if (!opt->population[idx].position) {
                fprintf(stderr, "Position for reef %d solution %d is null\n", i, j);
                return;
            }
            opt->population[idx].fitness = objective_function(opt->population[idx].position);
        }
    }
}

// Migration phase: exchange solutions between reefs
void migration_phase_cfo(Optimizer *opt) {
    if (!opt || !opt->population) {
        fprintf(stderr, "Invalid optimizer or population\n");
        return;
    }
    int solutions_per_reef = opt->population_size / NUM_REEFS;
    for (int i = 0; i < NUM_REEFS; i++) {
        for (int j = 0; j < NUM_REEFS; j++) {
            if (i != j) {
                int idx = i * solutions_per_reef + (rand() % solutions_per_reef);
                int idx_replace = j * solutions_per_reef + (rand() % solutions_per_reef);
                for (int k = 0; k < opt->dim; k++) {
                    opt->population[idx_replace].position[k] = opt->population[idx].position[k];
                }
                opt->population[idx_replace].fitness = INFINITY;  // Mark for re-evaluation
            }
        }
    }
    enforce_bound_constraints(opt);
}

// Local search phase: perturb solutions
void local_search_phase(Optimizer *opt) {
    if (!opt || !opt->population) {
        fprintf(stderr, "Invalid optimizer or population\n");
        return;
    }
    int solutions_per_reef = opt->population_size / NUM_REEFS;
    for (int i = 0; i < NUM_REEFS; i++) {
        for (int j = 0; j < solutions_per_reef; j++) {
            int idx = i * solutions_per_reef + j;
            if (!opt->population[idx].position) {
                fprintf(stderr, "Position for reef %d solution %d is null\n", i, j);
                return;
            }
            for (int k = 0; k < opt->dim; k++) {
                // Uniform perturbation instead of Gaussian for simplicity
                double perturbation = ALPHA * (2.0 * rand_double(0.0, 1.0) - 1.0);
                opt->population[idx].position[k] += perturbation;
            }
            opt->population[idx].fitness = INFINITY;  // Mark for re-evaluation
        }
    }
    enforce_bound_constraints(opt);
}

// Main optimization function
void CRO_optimize(Optimizer *opt, double (*objective_function)(double *)) {
    if (!opt || !objective_function) {
        fprintf(stderr, "Invalid optimizer or objective function\n");
        return;
    }

    // Seed random number generator
    srand((unsigned int)time(NULL));

    // Initialize reefs
    initialize_reefs(opt);
    if (!opt->population) {
        fprintf(stderr, "Initialization failed\n");
        return;
    }

    for (int iter = 0; iter < opt->max_iter; iter++) {
        // Evaluate fitness
        evaluate_reefs(opt, objective_function);

        // Find best solution across all reefs
        double min_fitness = INFINITY;
        int best_reef_idx = 0;
        int best_solution_idx = 0;
        int solutions_per_reef = opt->population_size / NUM_REEFS;
        for (int i = 0; i < NUM_REEFS; i++) {
            for (int j = 0; j < solutions_per_reef; j++) {
                int idx = i * solutions_per_reef + j;
                if (opt->population[idx].fitness < min_fitness) {
                    min_fitness = opt->population[idx].fitness;
                    best_reef_idx = i;
                    best_solution_idx = j;
                }
            }
        }
        if (min_fitness < opt->best_solution.fitness) {
            opt->best_solution.fitness = min_fitness;
            if (!opt->best_solution.position) {
                fprintf(stderr, "Best solution position is null\n");
                return;
            }
            int best_idx = best_reef_idx * solutions_per_reef + best_solution_idx;
            for (int k = 0; k < opt->dim; k++) {
                opt->best_solution.position[k] = opt->population[best_idx].position[k];
            }
        }

        // Migration phase
        migration_phase_cfo(opt);

        // Local search phase
        local_search_phase(opt);

        printf("Iteration %d: Best Value = %f\n", iter + 1, opt->best_solution.fitness);
    }
}
